### PR TITLE
[mlir] Add unit test for RankedTensorType wrapper class example.

### PR DESCRIPTION
**Pull request by jpienaar**
Add example as unit test for creating a wrapper type/view for RankedTensorType with encoding. This view provides a more restricted & typed API while it allows one to avoid repeated casting queries and accessing the encoding directly.

For users with more advance encodings, the expectation would be a separate attribute type, but here just StringAttr is used.

### CODE REVIEW

**Review by joker-eph — APPROVED** [review]

File: mlir/unittests/IR/ShapedTypeTest.cpp [hunk 1]
```diff
@@ -226,4 +227,63 @@ TEST(ShapedTypeTest, RankedTensorTypeBuilder) {
...
+  }
+
+  static bool classof(Type type);
+};
+
+bool TensorWithString::classof(Type type) {
+  if (auto rt = mlir::dyn_cast_or_null<RankedTensorType>(type))
+    return mlir::isa_and_present<StringAttr>(rt.getEncoding());
+  return false;
+}
```
Comment (River707) [inline mlir/unittests/IR/ShapedTypeTest.cpp:255]: Why is this the only method defined out-of-line?

File: mlir/unittests/IR/ShapedTypeTest.cpp [hunk 2]
```diff
@@ -226,4 +227,63 @@ TEST(ShapedTypeTest, RankedTensorTypeBuilder) {
...
+
+  static TensorWithString get(ArrayRef<int64_t> shape, Type elementType,
+                              StringRef name) {
+    return mlir::cast<TensorWithString>(RankedTensorType::get(
+        shape, elementType, StringAttr::get(elementType.getContext(), name)));
+  }
+
+  StringRef getName() const {
+    if (Attribute enc = getEncoding())
+      return mlir::cast<StringAttr>(enc).getValue();
```
Comment (River707) [inline mlir/unittests/IR/ShapedTypeTest.cpp:244]: If we're going to be adding these encoding wrappers, should we consider a templatized get encoding that does the cast/dyn_cast to the wrapped type?

**Review by River707 — APPROVED** [review]

File: mlir/unittests/IR/ShapedTypeTest.cpp (hunk 2 above)
Comment (jpienaar) [reply mlir/unittests/IR/ShapedTypeTest.cpp:244]: That's a good idea yes. Like "getEncodingAs<template>" ?

File: mlir/unittests/IR/ShapedTypeTest.cpp (hunk 1 above)
Comment (jpienaar) [reply mlir/unittests/IR/ShapedTypeTest.cpp:255]: Made inline for simplicity (no good reason here).

**Pull request merged**